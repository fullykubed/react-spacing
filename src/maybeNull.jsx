import React from 'react';

export function maybeNull({
  Component,
  ifRender,
  ifNull
}){
  if(typeof Component !== "function"){
    throw new Error('Component must be a React comoonent!');
  }
  const renderComponent = (result) =>
    result === null ? (ifNull ? ifNull() : null) : (ifRender ? ifRender(result) : null);
  if(Component.prototype && Component.prototype.render){
    return class extends Component {
      render() {
        return renderComponent(super.render())
      }
    }
  }else{
    return (props) => renderComponent(Component(props));
  }
}