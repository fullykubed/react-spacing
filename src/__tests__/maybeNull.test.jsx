import React from 'react';
import {mount} from 'enzyme';

import {maybeNull} from "../maybeNull";

const partialConfig = {
  ifRender: (children) => (
    <div id={"rendered"}>
      {children}
    </div>
  ),
  ifNull: () => (<div id="notRendered" />)
};

describe('maybeNull', () => {

  describe('functional components', () => {

    function NullComponent() {
      return null;
    }

    function NonNullComponent() {
      return <div/>;
    }

    it('renders ifRender' , () => {
      const WC = maybeNull({
        Component: NonNullComponent,
        ...partialConfig
      });
      expect(mount(<WC testProp={"jack"}/>)).toMatchSnapshot();
    });

    it('renders ifNull' , () => {
      const WC = maybeNull({
        Component: NullComponent,
        ...partialConfig
      });
      expect(mount(<WC/>)).toMatchSnapshot();
    });

    it('renders nothing' , () => {
      const WC = maybeNull({
        Component: NonNullComponent,
      });
      expect(mount(<WC/>)).toMatchSnapshot();
    });

    it('renders nothing 2' , () => {
      const WC = maybeNull({
        Component: NullComponent,
      });
      expect(mount(<WC/>)).toMatchSnapshot();
    });

  });

  describe('class components', () => {

    class NullComponent extends React.Component {
      render(){
        return null;
      }
    }

    class NonNullComponent extends React.Component {
      render(){
        return <div/>;
      }
    }

    it('renders ifRender' , () => {
      const WC = maybeNull({
        Component: NonNullComponent,
        ...partialConfig
      });
      expect(mount(<WC testProp={"jack"}/>)).toMatchSnapshot();
    });

    it('renders ifNull' , () => {
      const WC = maybeNull({
        Component: NullComponent,
        ...partialConfig
      });
      expect(mount(<WC/>)).toMatchSnapshot();
    });

    it('renders nothing' , () => {
      const WC = maybeNull({
        Component: NonNullComponent,
      });
      expect(mount(<WC/>)).toMatchSnapshot();
    });

    it('renders nothing 2' , () => {
      const WC = maybeNull({
        Component: NullComponent,
      });
      expect(mount(<WC/>)).toMatchSnapshot();
    });

  });


});