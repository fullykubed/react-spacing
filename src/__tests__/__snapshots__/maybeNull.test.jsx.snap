// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maybeNull class components renders ifNull 1`] = `
<_class>
  <div
    id="notRendered"
  />
</_class>
`;

exports[`maybeNull class components renders ifRender 1`] = `
<_class
  testProp="jack"
>
  <div
    id="rendered"
  >
    <div />
  </div>
</_class>
`;

exports[`maybeNull class components renders nothing 1`] = `<_class />`;

exports[`maybeNull class components renders nothing 2 1`] = `<_class />`;

exports[`maybeNull functional components renders ifNull 1`] = `
<bound >
  <div
    id="notRendered"
  />
</bound >
`;

exports[`maybeNull functional components renders ifRender 1`] = `
<bound 
  testProp="jack"
>
  <div
    id="rendered"
  >
    <div />
  </div>
</bound >
`;

exports[`maybeNull functional components renders nothing 1`] = `<bound  />`;

exports[`maybeNull functional components renders nothing 2 1`] = `<bound  />`;
